lib/ansible/modules/cloud/digital_ocean/digital_ocean_droplet.py

```diff
@@ -0,0 +1,308 @@
+    region: sfo1
+    image: ubuntu-16-04-x64
+    wait_timeout: 500
+'''
+
+
+RETURN = '''
+# Digital Ocean API info https://developers.digitalocean.com/documentation/v2/#droplets
+data:
+    description: a DigitalOcean Droplet
+    returned: changed
+    type: dict
+    sample: {
+        "droplet": {
+            "id": 3164494,
+            "name": "example.com",
+            "memory": 512,
+            "vcpus": 1,
+            "disk": 20,
+            "locked": true,
+            "status": "new",
+            "kernel": {
+                "id": 2233,
+                "name": "Ubuntu 14.04 x64 vmlinuz-3.13.0-37-generic",
+                "version": "3.13.0-37-generic"
+            },
+            "created_at": "2014-11-14T16:36:31Z",
+            "features": ["virtio"],
+            "backup_ids": [],
+            "snapshot_ids": [],
+            "image": {},
+            "volume_ids": [],
+            "size": {},
+            "size_slug": "512mb",
+            "networks": {},
+            "region": {},
+            "tags": ["web"]
+        }
+    }
+'''
+
+import time
+import traceback
+from ansible.module_utils._text import to_native
+from ansible.module_utils.basic import AnsibleModule, env_fallback
+from ansible.module_utils.digital_ocean import DigitalOceanHelper
+
+
+class DODroplet(object):
+    def __init__(self, module):
+        self.rest = DigitalOceanHelper(module)
+        self.module = module
+        self.wait = self.module.params.pop('wait', True)
+        self.wait_timeout = self.module.params.pop('wait_timeout', 120)
+        self.unique_name = self.module.params.pop('unique_name', False)
+        # pop the oauth token so we don't include it in the POST data
+        self.module.params.pop('oauth_token')
+
+    def get_by_id(self, droplet_id):
+        if not droplet_id:
+            return None
+        response = self.rest.get('droplets/{0}'.format(droplet_id))
+        json_data = response.json
+        if response.status_code == 200:
+            return json_data
+        return None
+
+    def get_by_name(self, droplet_name):
+        if not droplet_name:
+            return None
+        page = 1
+        while page is not None:
+            response = self.rest.get('droplets?page={0}'.format(page))
+            json_data = response.json
+            if response.status_code == 200:
+                for droplet in json_data['droplets']:
+                    if droplet['name'] == droplet_name:
+                        return {'droplet': droplet}
+                if 'links' in json_data and 'pages' in json_data['links'] and 'next' in json_data['links']['pages']:
+                    page += 1
+                else:
+                    page = None
+        return None
+
+    def get_droplet(self):
+        json_data = self.get_by_id(self.module.params['id'])
+        if not json_data and self.unique_name:
+            json_data = self.get_by_name(self.module.params['name'])
+        return json_data
+
+    def create(self):
```

Comment:
@gurch101 here is suggested change to the create method based on the feedback from other individuals. We do need to include the ip addresses at the top level of the response so they can be easily passed into additional tasks or facts. Let me know what you think of this method change..

```python
    def create(self):
        json_data = self.get_droplet()
        droplet_data = None
        if json_data:
            droplet_data = self.get_addresses(json_data)
            self.module.exit_json(changed=False, data=droplet_data)
        response = self.rest.post('droplets', data=self.module.params)
        json_data = response.json
        if response.status_code == 422 and json_data['message'] == 'Region is not available':
            self.module.fail_json(chnaged=False, msg=json_data['message'])
        if self.wait:
            json_data = self.ensure_power_on(json_data['droplet']['id'])
            droplet_data = self.get_addresses(json_data)
        self.module.exit_json(changed=True, data=droplet_data)
```

The address method is used to pull out the different ip addresses DO sends back in the response.

```python
    def get_addresses(self, data):
        """
         Expose IP addresses as their own property allowing users extend to additional tasks
        """
        _data = data
        for k, v in data.items():
            setattr(self, k, v)
        networks = _data['droplet']['networks']
        for network in networks.get('v4', []):
            if network['type'] == 'public':
                _data['ip_address'] = network['ip_address']
            else:
                _data['private_ipv4_address'] = network['ip_address']
        for network in networks.get('v6', []):
            if network['type'] == 'public':
                _data['ipv6_address'] = network['ip_address']
            else:
                _data['private_ipv6_address'] = network['ip_address']
        return _data
```